src/swarm/neo/util/Batch.d

```diff
@@ -148,9 +148,78 @@ struct BatchWriter
     }
 }
 
-struct BatchReader
+scope class BatchReader ( T ... )
 {
+    private Const!(void)[] batch;
+    private Const!(void)[] remaining;
+
+    public this ( in void[] batch )
+    {
+        this.batch = batch;
+        this.remaining = this.batch;
+    }
+
+    // compress_buf must remain accessible for the life of this instance
+    public this ( Lzo lzo, in void[] batch, ref void[] compress_buf )
+    {
+        // Read uncompressed length from first size_t.sizeof bytes.
+        auto uncompressed_len = *(cast(size_t*)(batch.ptr));
+        compress_buf.length = uncompressed_len;
+
+        // Decompress into this.batch.
+        auto src = batch[size_t.sizeof .. $];
+        lzo.uncompress(src, compress_buf);
 
+        // Construct as normal, extracting from the decompressed data.
+        this(compress_buf);
+    }
+
+    public int opApply ( int delegate ( ref T ) dg )
+    {
+        T items;
+        while ( this.remaining.length )
+        {
+            foreach ( ref item; items )
+                this.extractItem(item);
+
+            if ( auto r = dg(items) )
+                return r;
+        }
+
+        return 0;
+    }
+
+    private void extractItem ( T ) ( ref T item )
+    {
+        static if ( isDynamicArrayType!(T) )
+        {
+            static assert(rankOfArray!(T) == 1,
+                "Only 1d arrays and value types are supported, not "
+                ~ T.stringof);
+            auto len = this.extractBytes(size_t.sizeof);
+            item = cast(T)this.extractBytes(*(cast(size_t*)len.ptr));
+        }
+        else
+        {
+            static assert(!hasIndirections!(T),
+                "Only 1d arrays and value types are supported, not "
+                ~ T.stringof);
+            auto slice = (cast(void*)&item)[0..T.sizeof];
+            slice[] = this.extractBytes(T.sizeof)[];
+        }
+    }
+
+    private Const!(void)[] extractBytes ( size_t bytes )
+    in
+    {
+        assert(this.remaining.length >= bytes);
```

Comment:
This assertion is pointless because it does the same as the array bounds checks in the function body. This condition needs to be checked with `enforce`, though, because `this.remaining` contains external input data.

Reply:
Ah yes, I wondered if someone would notice that. I included it as an in contract so that it'll be translated into a `verify` when we convert swarm. You're right, though, `enforce` also makes sense.